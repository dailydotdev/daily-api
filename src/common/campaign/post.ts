import { ValidationError } from 'apollo-server-errors';
import { AuthContext } from '../../Context';
import {
  ArticlePost,
  CampaignPost,
  CampaignState,
  CampaignType,
  Post,
  PostType,
  type ConnectionManager,
  type FreeformPost,
  type SharePost,
} from '../../entity';
import { getPostPermalink } from '../../schema/posts';
import {
  type GetCampaignListResponse,
  type GetCampaignResponse,
} from '../../integrations/skadi';
import type { Connection } from 'graphql-relay';
import { In } from 'typeorm';
import { mapCloudinaryUrl } from '../cloudinary';
import { pickImageUrl } from '../post';
import { NotFoundError } from '../../errors';
import { debeziumTimeToDate, systemUser, updateFlagsStatement } from '../utils';
import { getDiscussionLink } from '../links';
import { skadiApiClientV1 } from '../../integrations/skadi/api/v1/clients';
import { largeNumberFormat } from '../devcard';
import { formatMailDate, addNotificationEmailUtm } from '../mailing';
import { truncatePostToTweet } from '../twitter';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import { usdToCores } from '../number';

import {
  startCampaignTransferCores,
  stopCampaignTransferCores,
  type StartCampaignMutationArgs,
  type StopCampaignProps,
} from './common';
import { randomUUID } from 'crypto';
import {
  UserTransaction,
  UserTransactionProcessor,
  UserTransactionStatus,
  UserTransactionType,
} from '../../entity/user/UserTransaction';
import { addDays } from 'date-fns';
import { skadiApiClientV2 } from '../../integrations/skadi/api/v2/clients';

export interface GQLPromotedPost
  extends Omit<
    GetCampaignResponse,
    'spend' | 'budget' | 'startedAt' | 'endedAt'
  > {
  spend: number;
  budget: number;
  startedAt: Date;
  endedAt: Date;
}

export const validatePostBoostPermissions = async (
  ctx: AuthContext,
  postId: string,
): Promise<Pick<Post, 'id' | 'flags'>> => {
  const { userId } = ctx;

  return ctx.con.getRepository(Post).findOneOrFail({
    select: ['id', 'flags'],
    where: [
      { id: postId, authorId: userId },
      { id: postId, scoutId: userId },
    ],
  });
};

export const checkPostAlreadyBoosted = (post: Pick<Post, 'flags'>): void => {
  if (!!post.flags?.campaignId) {
    throw new ValidationError('Post is already boosted');
  }
};

interface CampaignBoostedPost
  extends Pick<Post, 'id' | 'shortId' | 'title' | 'slug'> {
  image: string;
  permalink: string;
  engagements: number;
  commentsPermalink?: string;
}

export interface GQLBoostedPost {
  campaign: GQLPromotedPost;
  post: CampaignBoostedPost;
}

interface GetBoostedPost extends CampaignBoostedPost {
  type: PostType;
  sharedTitle?: string;
  sharedImage?: string;
  views: number;
  upvotes: number;
  comments: number;
}

const getBoostedPostBuilder = (con: ConnectionManager, alias = 'p1') =>
  con
    .getRepository(Post)
    .createQueryBuilder(alias)
    .select(`"${alias}".id`, 'id')
    .addSelect(`"${alias}"."shortId"`, 'shortId')
    .addSelect(`"${alias}".slug`, 'slug')
    .addSelect(`"${alias}".image`, 'image')
    .addSelect(`"${alias}".title`, 'title')
    .addSelect(`"${alias}".type`, 'type')
    .addSelect(`"${alias}".upvotes::int`, 'upvotes')
    .addSelect(`"${alias}".comments::int`, 'comments')
    .addSelect(`"${alias}".views::int`, 'views')
    .addSelect('p2.title', 'sharedTitle')
    .addSelect('p2.image', 'sharedImage')
    .leftJoin(Post, 'p2', `"${alias}"."sharedPostId" = p2.id`);

export const getBoostedPost = async (
  con: ConnectionManager,
  id: string,
): Promise<GetBoostedPost> => {
  const result = await getBoostedPostBuilder(con).where({ id }).getRawOne();

  if (!result) {
    throw new NotFoundError('Post does not exist');
  }

  return result;
};

export const getFormattedBoostedPost = (
  post: GetBoostedPost,
): GQLBoostedPost['post'] => {
  const { id, shortId, sharedImage, sharedTitle, slug } = post;
  let image: string | undefined = post.image;
  let title = post.title;

  if (post.type === PostType.Share) {
    image = sharedImage;
    title = title || sharedTitle;
  }

  return {
    id,
    slug,
    shortId,
    title,
    image: mapCloudinaryUrl(image) ?? pickImageUrl({ createdAt: new Date() }),
    permalink: getPostPermalink({ shortId }),
    commentsPermalink: post.slug ? getDiscussionLink(post.slug) : undefined,
    engagements: post.comments + post.upvotes + post.views,
  };
};

export const getFormattedCampaign = ({
  spend,
  budget,
  startedAt,
  endedAt,
  ...campaign
}: GetCampaignResponse): GQLPromotedPost => ({
  ...campaign,
  spend: usdToCores(parseFloat(spend)),
  budget: usdToCores(parseFloat(budget)),
  startedAt: debeziumTimeToDate(startedAt),
  endedAt: debeziumTimeToDate(endedAt),
});

export interface BoostedPostStats
  extends Pick<GetCampaignListResponse, 'clicks' | 'impressions' | 'users'> {
  engagements: number;
  totalSpend: number;
}

export interface BoostedPostConnection extends Connection<GQLBoostedPost> {
  stats?: BoostedPostStats;
}

export const consolidateCampaignsWithPosts = async (
  campaigns: GetCampaignResponse[],
  con: ConnectionManager,
): Promise<GQLBoostedPost[]> => {
  const ids = campaigns.map(({ postId }) => postId);
  const builder = getBoostedPostBuilder(con);
  const postAlias = 'p1';
  const posts = await builder
    .where(`"${postAlias}".id IN (:...ids)`, { ids })
    .getRawMany<GetBoostedPost>();
  const mapped = posts.reduce(
    (map, post) => ({ ...map, [post.id]: post }),
    {} as Record<string, GetBoostedPost>,
  );

  return campaigns.map((campaign) => ({
    campaign: getFormattedCampaign(campaign),
    post: getFormattedBoostedPost(mapped[campaign.postId]),
  }));
};

interface TotalEngagements
  extends Pick<Post, 'views' | 'comments' | 'upvotes'> {}

export const getTotalEngagements = async (
  con: ConnectionManager,
  postIds: string[],
): Promise<number> => {
  const builder = con.getRepository(Post).createQueryBuilder();

  const engagements = await builder
    .select('SUM(upvotes)::int', 'upvotes')
    .addSelect('SUM(comments)::int', 'comments')
    .addSelect('SUM(views)::int', 'views')
    .where({ id: In(postIds) })
    .getRawOne<TotalEngagements>();

  return (
    (engagements?.upvotes || 0) +
    (engagements?.comments || 0) +
    (engagements?.views || 0)
  );
};

export const generateBoostEmailUpdate: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await skadiApiClientV1.getCampaignById({
    campaignId: notification.referenceId!,
    userId: user.id,
  });

  if (!campaign) {
    return null;
  }

  const post = await con.getRepository(Post).findOne({
    where: { id: campaign.postId },
  });

  if (!post) {
    return null;
  }

  const sharedPost = await (post.type === PostType.Share
    ? con.getRepository(ArticlePost).findOne({
        where: { id: (post as SharePost).sharedPostId },
        select: ['title', 'image', 'slug'],
      })
    : Promise.resolve(null));

  const title = truncatePostToTweet(post || sharedPost);
  const engagement = post.views + post.upvotes + post.comments;

  return {
    start_date: formatMailDate(debeziumTimeToDate(campaign.startedAt)),
    end_date: formatMailDate(debeziumTimeToDate(campaign.endedAt)),
    impressions: largeNumberFormat(campaign.impressions),
    clicks: largeNumberFormat(campaign.clicks),
    engagement: largeNumberFormat(engagement),
    post_link: getDiscussionLink(post.slug),
    analytics_link: addNotificationEmailUtm(
      notification.targetUrl,
      notification.type,
    ),
    post_image: sharedPost?.image || (post as FreeformPost).image,
    post_title: title,
  };
};

export const getAdjustedReach = (value: number) => {
  // We do plus-minus 8% of the generated value
  const difference = Math.floor(value * 0.08);
  const min = Math.max(value - difference, 0);
  const estimatedReach = {
    min,
    max: Math.max(value + difference, min),
  };

  return estimatedReach;
};

export const startCampaignPost = async (props: StartCampaignMutationArgs) => {
  const { ctx, args } = props;
  const { value: postId } = args;
  const post = await validatePostBoostPermissions(ctx, postId);
  checkPostAlreadyBoosted(post);

  const request = await ctx.con.transaction(async (manager) => {
    const id = randomUUID();
    const creativeId = randomUUID();
    const { budget, duration } = args;
    const total = budget * duration;
    const userId = ctx.userId;
    const endedAt = addDays(new Date(), duration);

    const campaign = await manager.getRepository(CampaignPost).save(
      manager.getRepository(CampaignPost).create({
        id,
        creativeId,
        flags: {
          budget: total,
          spend: 0,
          users: 0,
          clicks: 0,
          impressions: 0,
        },
        userId,
        referenceId: postId,
        state: CampaignState.Active,
        endedAt,
        postId,
        type: CampaignType.Post,
      }),
    );

    const campaignId = campaign.id;
    const tags = await getPostTags(manager, postId);

    await skadiApiClientV2.startCampaign(campaign, tags);

    await manager
      .getRepository(Post)
      .update(
        { id: postId },
        { flags: updateFlagsStatement<Post>({ campaignId: id }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: systemUser.id,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: userId,
        value: total,
        valueIncFees: 0,
        fee: 0,
        request: ctx.requestMeta,
        flags: { note: `Post Boost started` },
        referenceId: campaignId,
        referenceType: UserTransactionType.PostBoost,
      }),
    );

    return await startCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return request.transaction;
};

export const stopCampaignPost = async ({
  campaign,
  ctx,
}: StopCampaignProps) => {
  const { id: campaignId, userId, referenceId } = campaign;

  const { budget } = await skadiApiClientV2.cancelCampaign({
    campaignId,
    userId,
  });

  const result = await ctx.con.transaction(async (manager) => {
    const toRefund = parseFloat(budget);

    await manager
      .getRepository(Post)
      .update(
        { id: referenceId },
        { flags: updateFlagsStatement<Post>({ campaignId: null }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: userId,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: systemUser.id,
        value: usdToCores(toRefund),
        valueIncFees: 0,
        fee: 0,
        flags: { note: `Post Boost refund` },
        referenceId: campaignId,
        referenceType: UserTransactionType.PostBoost,
      }),
    );

    return await stopCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return result.transaction;
};

export const getPostTags = async (con: ConnectionManager, postId: string) => {
  const builder = con.getRepository(Post).createQueryBuilder('p1');
  const subquery = builder
    .subQuery()
    .select(`COALESCE(p2."tagsStr", '')`)
    .from(Post, 'p2')
    .where('p2.id = p1."sharedPostId"')
    .getQuery();

  const result = await builder
    .select(`COALESCE(p1."tagsStr", '')`, 'tagsStr')
    .addSelect(`(${subquery})`, 'sharedTagsStr')
    .where('p1.id = :id', { id: postId })
    .getRawOne<{ tagsStr: string; sharedTagsStr: string }>();

  const tags1 = (result?.tagsStr ?? '').split(',');
  const tags2 = (result?.sharedTagsStr ?? '').split(',');
  const list = [...tags1, ...tags2].filter((tag) => tag.trim().length > 0);

  return Array.from(new Set(list));
};
