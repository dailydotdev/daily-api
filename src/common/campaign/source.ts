import {
  CampaignSource,
  CampaignState,
  CampaignType,
  Post,
  Source,
  type ConnectionManager,
} from '../../entity';

import type { AuthContext } from '../../Context';
import {
  startCampaignTransferCores,
  stopCampaignTransferCores,
  type StartCampaignMutationArgs,
  type StopCampaignProps,
} from './common';
import {
  ensureSourcePermissions,
  SourcePermissions,
} from '../../schema/sources';
import { ValidationError } from 'apollo-server-errors';
import { randomUUID } from 'crypto';
import { addDays } from 'date-fns';
import {
  UserTransaction,
  UserTransactionProcessor,
  UserTransactionStatus,
  UserTransactionType,
} from '../../entity/user/UserTransaction';
import { usdToCores } from '../number';
import { systemUser, updateFlagsStatement } from '../utils';
import { skadiApiClientV2 } from '../../integrations/skadi/api/v2/clients';
import { NotificationIcon } from '../../notifications/icons';
import { notificationsLink } from '../links';
import type { NotificationBuilder } from '../../notifications/builder';
import { formatMailDate, addNotificationEmailUtm } from '../mailing';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import type { NotificationCampaignSourceContext } from '../../notifications';

export const validateSquadBoostPermissions = async (
  ctx: AuthContext,
  sourceId: string,
) => {
  const source = await ensureSourcePermissions(
    ctx,
    sourceId,
    SourcePermissions.BoostSquad,
  );

  if (source.flags.campaignId) {
    throw new ValidationError('Source already has a campaign');
  }

  return source;
};

export const getSourceTags = async (
  con: ConnectionManager,
  sourceId: string,
): Promise<string[]> => {
  const result = await con.getRepository(Post).query<{ tag: string }[]>(
    `
      WITH recent_posts AS (
        SELECT id, "sharedPostId"
        FROM post
        WHERE "sourceId" = $1
        ORDER BY "createdAt" DESC
        LIMIT 30
      )
      SELECT DISTINCT ps.keyword AS tag
      FROM post_keyword ps
      INNER JOIN recent_posts rp
        ON ps."postId" = rp.id
        OR ps."postId" = rp."sharedPostId"
      WHERE ps.status = 'allow'
      LIMIT 30;
    `,
    [sourceId],
  );

  return result.map(({ tag }) => tag.trim()).filter(Boolean);
};

export const startCampaignSource = async (props: StartCampaignMutationArgs) => {
  const { ctx, args } = props;
  const { value } = args;
  const source = await validateSquadBoostPermissions(ctx, value);

  const request = await ctx.con.transaction(async (manager) => {
    const id = randomUUID();
    const { budget, duration } = args;
    const total = budget * duration;
    const userId = ctx.userId;
    const endedAt = addDays(new Date(), duration);

    const campaign = await manager.getRepository(CampaignSource).save(
      manager.getRepository(CampaignSource).create({
        id,
        flags: {
          budget: total,
          spend: 0,
          users: 0,
          clicks: 0,
          impressions: 0,
        },
        userId,
        referenceId: source.id,
        state: CampaignState.Active,
        endedAt,
        sourceId: source.id,
        type: CampaignType.Squad,
      }),
    );

    const campaignId = campaign.id;
    const last30tags = await getSourceTags(manager, source.id);
    const finalTags = last30tags.length > 3 ? last30tags : []; // when it is 3 or below, we set global targeting

    await skadiApiClientV2.startCampaign(campaign, finalTags);

    await manager
      .getRepository(Source)
      .update(
        { id: source.id },
        { flags: updateFlagsStatement<Source>({ campaignId: id }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: systemUser.id,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: userId,
        value: total,
        valueIncFees: 0,
        fee: 0,
        request: ctx.requestMeta,
        flags: { note: `Squad Boost started` },
        referenceId: campaignId,
        referenceType: UserTransactionType.SquadBoost,
      }),
    );

    return await startCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return request.transaction;
};

export const stopCampaignSource = async ({
  campaign,
  ctx,
}: StopCampaignProps) => {
  const { id: campaignId, userId, referenceId } = campaign;

  const { budget } = await skadiApiClientV2.cancelCampaign({
    campaignId,
    userId,
  });

  const result = await ctx.con.transaction(async (manager) => {
    const toRefund = parseFloat(budget);

    await manager
      .getRepository(CampaignSource)
      .update({ id: campaignId }, { state: CampaignState.Cancelled });

    await manager
      .getRepository(Source)
      .update(
        { id: referenceId },
        { flags: updateFlagsStatement<Source>({ campaignId: null }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: userId,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: systemUser.id,
        value: usdToCores(toRefund),
        valueIncFees: 0,
        fee: 0,
        flags: { note: `Squad Boost refund` },
        referenceId: campaignId,
        referenceType: UserTransactionType.SquadBoost,
      }),
    );

    return await stopCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return result.transaction;
};

export const generateCampaignSquadNotification = (
  builder: NotificationBuilder,
  ctx: NotificationCampaignSourceContext,
) => {
  const { campaign, source, event, user } = ctx;

  return builder
    .icon(NotificationIcon.DailyDev)
    .referenceCampaign(ctx)
    .targetUrl(notificationsLink)
    .setTargetUrlParameter([['c_id', campaign.id]])
    .uniqueKey(`${campaign.id}-${user.id}-${event}`)
    .avatarSource(source);
};

export const generateCampaignSquadEmail: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await con
    .getRepository(CampaignSource)
    .findOneBy({ id: notification.referenceId });

  if (!campaign) {
    return null;
  }

  const source = await con
    .getRepository(Source)
    .findOneByOrFail({ id: campaign.sourceId });

  return {
    start_date: formatMailDate(campaign.createdAt),
    end_date: formatMailDate(campaign.endedAt),
    analytics_link: addNotificationEmailUtm(
      notification.targetUrl,
      notification.type,
    ),
    source_image: source.image,
    source_handle: source.handle,
    source_name: source.name,
  };
};
