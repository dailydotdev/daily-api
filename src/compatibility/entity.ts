import { SourceRequest } from '../entity';
import { ChangeObject } from '../types';

export interface LegacySourceRequest {
  id: string;
  url: string;
  userId: string;
  userName?: string;
  userEmail?: string;
  approved?: boolean;
  closed?: boolean;
  pubId?: string;
  pubName?: string;
  pubImage?: string;
  pubTwitter?: string;
  pubRss?: string;
  createdAt: Date;
  reason?: string;
}

export const toLegacySourceRequest = (
  req: ChangeObject<SourceRequest> | SourceRequest,
): LegacySourceRequest => ({
  id: req.id,
  url: req.sourceUrl,
  userId: req.userId,
  userName: req.userName,
  userEmail: req.userEmail,
  approved: req.approved,
  closed: req.closed,
  pubId: req.sourceId,
  pubName: req.sourceName,
  pubImage: req.sourceImage,
  pubTwitter: req.sourceTwitter,
  pubRss: req.sourceFeed,
  createdAt: new Date(req.createdAt),
  reason: req.reason,
});
