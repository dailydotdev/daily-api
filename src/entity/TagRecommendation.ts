import { ViewEntity, ViewColumn, Index } from 'typeorm';

@ViewEntity({
  materialized: true,
  expression: `
    with keywords as (
        select pk."postId", pk.keyword
        from post_keyword pk
        join post p on p.id = pk."postId"
        where p."createdAt" >= (current_timestamp - interval '60 day')::date and pk.status = 'allow'
    ), totals as (
        select keyword, count(*) total
        from keywords
        group by 1
    ), filtered as (
        select k.*
        from keywords k
        join totals t on t.keyword = k.keyword
        where t.total >= 100
    ), pairs as (
        select k1.keyword "keywordX", k2.keyword "keywordY"
        from filtered k1
        join filtered k2 on k1."postId" = k2."postId" and k1.keyword != k2.keyword
    ), likelihood as (
        select "keywordX", "keywordY", count(*) * 1.0 / min(t.total) as probability
        from pairs p
        join totals t on p."keywordX" = t.keyword
        group by 1, 2
    )
    select * from likelihood
    where probability > 0.05
  `,
})
@Index('IDX_keywordX_keywordY_probability', [
  'keywordX',
  'keywordY',
  'probability',
])
export class TagRecommendation {
  @ViewColumn()
  keywordX: string;

  @ViewColumn()
  keywordY: string;

  @ViewColumn()
  probability: number;
}
