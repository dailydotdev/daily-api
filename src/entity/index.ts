export * from './user';
export * from './AdvancedSettings';
export * from './Alerts';
export * from './Banner';
export * from './Bookmark';
export * from './BookmarkList';
export * from './DevCard';
export * from './Feed';
export * from './FeedAdvancedSettings';
export * from './FeedSource';
export * from './FeedTag';
export * from './Integration';
export * from './posts';
export * from './PostReport';
export * from './PostTag';
export * from './Settings';
export * from './Source';
export * from './SourceDisplay';
export * from './SourceFeed';
export * from './SourceMember';
export * from './SourceRequest';
export * from './TagSegment';
export * from './View';
export * from './Comment';
export * from './Keyword';
export * from './PostKeyword';
export * from './CommentMention';
export * from './ReputationEvent';
export * from './Submission';
export * from './notifications';
export * from './Feature';
export * from './Invite';
export * from './ContentImage';
export * from './MarketingCta';
export * from './SquadPublicRequest';
