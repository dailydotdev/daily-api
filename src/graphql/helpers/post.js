const getFeedParams =
  ({ user, post }, args, rankBy, filters = {}, ignoreUserFilters = false, ip) => {
    const userId = user ? user.userId : null;

    return {
      fields: userId ? [...post.defaultUserFields, 'read'] : post.defaultAnonymousFields,
      filters: Object.assign({}, { before: args.latest }, filters),
      rankBy,
      userId,
      ignoreUserFilters,
      page: args.page,
      pageSize: args.pageSize,
      ip,
    };
  };

const assignType = type => x => Object.assign({}, x, { type });

export default {
  getFeedParams,
  assignType,
};
