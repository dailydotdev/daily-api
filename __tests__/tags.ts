import {
  disposeGraphQLTesting,
  GraphQLTestClient,
  GraphQLTestingState,
  initializeGraphQLTesting,
  MockContext,
  saveFixtures,
  testQueryError,
} from './helpers';
import { ArticlePost, Keyword, Post, PostKeyword, Source } from '../src/entity';
import {
  keywordsFixture,
  postRecommendedKeywordsFixture,
} from './fixture/keywords';
import { DataSource } from 'typeorm';
import createOrGetConnection from '../src/db';
import { postsFixture, postKeywordsFixture } from './fixture/post';
import { sourcesFixture } from './fixture/source';
import { TagRecommendation } from '../src/entity/TagRecommendation';
import { SubmissionFailErrorMessage } from '../src/errors';

let con: DataSource;
let state: GraphQLTestingState;
let client: GraphQLTestClient;
let loggedUser: string = null;

beforeAll(async () => {
  con = await createOrGetConnection();
  state = await initializeGraphQLTesting(
    () => new MockContext(con, loggedUser),
  );
  client = state.client;
});

beforeEach(async () => {
  loggedUser = null;

  await saveFixtures(con, Keyword, keywordsFixture);
});

afterAll(() => disposeGraphQLTesting(state));

describe('query tags', () => {
  const QUERY = `{
    tags {
      value
    }
  }`;

  it('should return all tags', async () => {
    const res = await client.query(QUERY);
    expect(res.data).toMatchObject({
      tags: [
        { value: 'development' },
        { value: 'fullstack' },
        { value: 'golang' },
        { value: 'rust' },
        { value: 'webdev' },
      ],
    });
  });
});

describe('query trendingTags', () => {
  const QUERY = `{
    trendingTags {
      name
    }
  }`;

  beforeEach(async () => {
    await saveFixtures(con, Source, sourcesFixture);
    let tags = 'tag1';
    await con.getRepository(Post).save(
      new Array(20).fill('tag').map((item, index) => {
        tags += `,tag${index + 1}`;
        return {
          id: `post_${index}`,
          shortId: `post_${index}`,
          title: `Post ${index}`,
          tagsStr: tags,
          upvotes: 10 + index,
          createdAt: new Date(),
          sourceId: 'a',
        };
      }),
    );
    await con.query(`REFRESH MATERIALIZED VIEW trending_post`);
    await con.query(`REFRESH MATERIALIZED VIEW trending_tag`);
  });

  it('should return most trending tags ordered by value', async () => {
    const res = await client.query(QUERY);
    expect(res.data).toMatchObject({
      trendingTags: [
        { name: 'tag19' },
        { name: 'tag18' },
        { name: 'tag17' },
        { name: 'tag16' },
        { name: 'tag15' },
        { name: 'tag14' },
        { name: 'tag13' },
        { name: 'tag12' },
        { name: 'tag11' },
        { name: 'tag10' },
      ],
    });
  });

  it('should return limit of 10 by default', async () => {
    const res = await client.query(QUERY);
    expect(res.data.trendingTags.length).toBe(10);
  });
});

describe('query popularTags', () => {
  const QUERY = `{
    popularTags {
      name
    }
  }`;

  beforeEach(async () => {
    await saveFixtures(con, Source, sourcesFixture);
    let tags = 'tag1';
    await con.getRepository(Post).save(
      new Array(20).fill('tag').map((item, index) => {
        tags += `,tag${index + 1}`;
        return {
          id: `post_${index}`,
          shortId: `post_${index}`,
          title: `Post ${index}`,
          tagsStr: tags,
          upvotes: 10 + index,
          createdAt: new Date(),
          sourceId: 'a',
        };
      }),
    );
    await con.query(`REFRESH MATERIALIZED VIEW popular_post`);
    await con.query(`REFRESH MATERIALIZED VIEW popular_tag`);
  });

  it('should return most popular tags ordered by value', async () => {
    const res = await client.query(QUERY);
    expect(res.data).toMatchObject({
      popularTags: [
        { name: 'tag10' },
        { name: 'tag9' },
        { name: 'tag8' },
        { name: 'tag7' },
        { name: 'tag6' },
        { name: 'tag5' },
        { name: 'tag4' },
        { name: 'tag3' },
        { name: 'tag2' },
        { name: 'tag1' },
      ],
    });
  });

  it('should return limit of 10 by default', async () => {
    const res = await client.query(QUERY);
    expect(res.data.popularTags.length).toBe(10);
  });
});

describe('query searchTags', () => {
  const QUERY = (query: string): string => `{
    searchTags(query: "${query}") {
      query
      hits {
        name
      }
    }
  }`;

  it('should search for tags and order by value', async () => {
    const res = await client.query(QUERY('dev'));
    expect(res.data).toMatchSnapshot();
  });

  it('should take into account keyword synonyms', async () => {
    const res = await client.query(QUERY('web-dev'));
    expect(res.data).toMatchSnapshot();
  });

  it('should return no results if query length is less then 2', async () => {
    const res = await client.query(QUERY('d'));
    expect(res.data).toMatchObject({
      searchTags: {
        query: 'd',
        hits: [],
      },
    });
  });

  it('should return no more then 100 results', async () => {
    await con.getRepository(Keyword).save(
      new Array(110).fill('tag').map((item, index) => ({
        value: item + index,
        occurances: 0,
        status: 'allow',
      })),
    );

    const res = await client.query(QUERY('tag'));

    expect(res.errors).toBeFalsy();
    expect(res.data.searchTags.hits.length).toBe(100);
  });
});

describe('query onboardingTags', () => {
  const QUERY = `{
    onboardingTags {
      hits {
        name
      }
    }
  }`;

  beforeEach(async () => {
    await saveFixtures(
      con,
      Keyword,
      keywordsFixture.map((item) => ({
        ...item,
        flags: {
          onboarding: true,
        },
      })),
    );
  });

  it('should return onboarding tags', async () => {
    const res = await client.query(QUERY);

    expect(res.errors).toBeFalsy();
    expect(res.data).toMatchObject({
      onboardingTags: {
        hits: [
          { name: 'development' },
          { name: 'fullstack' },
          { name: 'golang' },
          { name: 'pending' },
          { name: 'politics' },
          { name: 'rust' },
          { name: 'web-development' },
          { name: 'webdev' },
        ],
      },
    });
  });
});

describe('query recommendedTags', () => {
  const QUERY = `
    query recommendedTags($tags: [String]!, $excludedTags: [String]!) {
      recommendedTags(tags: $tags, excludedTags: $excludedTags) {
        hits {
          name
        }
      }
    }`;

  beforeEach(async () => {
    await saveFixtures(con, Source, sourcesFixture);
    await saveFixtures(con, ArticlePost, postsFixture);
    await saveFixtures(
      con,
      Keyword,
      postRecommendedKeywordsFixture.map((item) => ({
        ...item,
        flags: {
          onboarding: true,
        },
      })),
    );
    await saveFixtures(
      con,
      PostKeyword,
      postKeywordsFixture.map((item) => ({
        ...item,
        status: 'allow',
      })),
    );

    const materializedViewName =
      con.getRepository(TagRecommendation).metadata.tableName;
    await con.query(`REFRESH MATERIALIZED VIEW ${materializedViewName}`);
  });

  it('should return recommended tags', async () => {
    const res = await client.query(QUERY, {
      variables: {
        tags: ['javascript'],
        excludedTags: [],
      },
    });

    expect(res.errors).toBeFalsy();
    expect(res.data).toMatchObject({
      recommendedTags: {
        hits: [
          { name: 'backend' },
          { name: 'data' },
          { name: 'html' },
          { name: 'webdev' },
        ],
      },
    });
  });

  it('should not include origin tags in recommended tags', async () => {
    const originTags = ['javascript', 'webdev'];

    const res = await client.query(QUERY, {
      variables: {
        tags: originTags,
        excludedTags: [],
      },
    });

    expect(res.errors).toBeFalsy();
    originTags.forEach((tag) => {
      expect(
        res.data.recommendedTags.hits.find(
          (hit: { name: string }) => hit.name === tag,
        ),
      ).toBeFalsy();
    });
  });

  it('should not return excluded tags in recommended tags', async () => {
    const res = await client.query(QUERY, {
      variables: {
        tags: ['javascript'],
        excludedTags: ['html', 'data'],
      },
    });

    expect(res.errors).toBeFalsy();
    expect(res.data).toMatchObject({
      recommendedTags: {
        hits: [{ name: 'backend' }, { name: 'webdev' }],
      },
    });
  });

  it('should return empty array if no tags are provided', async () => {
    const res = await client.query(QUERY, {
      variables: {
        tags: [],
        excludedTags: [],
      },
    });

    expect(res.errors).toBeFalsy();
    expect(res.data).toMatchObject({
      recommendedTags: {
        hits: [],
      },
    });
  });

  it('should throw validation error if more then 1000 tags is included', async () => {
    await testQueryError(
      client,
      {
        query: QUERY,
        variables: {
          tags: new Array(800).fill('tag').map((item, index) => item + index),
          excludedTags: new Array(500)
            .fill('excTag')
            .map((item, index) => item + index),
        },
      },
      (errors) => {
        expect(errors[0].extensions?.code).toEqual('GRAPHQL_VALIDATION_FAILED');
        expect(errors[0].message).toEqual(
          SubmissionFailErrorMessage.ONBOARDING_TAG_LIMIT_REACHED,
        );
      },
    );
  });
});
