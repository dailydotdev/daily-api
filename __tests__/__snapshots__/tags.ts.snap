// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query searchTags should search for tags and order by value 1`] = `
Object {
  "searchTags": Object {
    "hits": Array [
      Object {
        "name": "development",
      },
      Object {
        "name": "webdev",
      },
    ],
    "query": "dev",
  },
}
`;

exports[`query searchTags should take into account keyword synonyms 1`] = `
Object {
  "searchTags": Object {
    "hits": Array [],
    "query": "web-dev",
  },
}
`;
